h5py/h5t.pyx

```diff
@@ -1608,9 +1610,16 @@ cpdef TypeID py_create(object dtype_in, bint logical=0, bint aligned=0):
         of kind "O" representing a string, it would return an HDF5 variable-
         length string type.
     """
-    cdef dtype dt = dtype(dtype_in)
-    cdef char kind = dt.kind
+    cdef:
+        cnp.dtype dt
+        char kind
 
+    if isinstance(dtype_in, np.dtype):
+        dt = np.dtype(dtype_in)
+    else:
+        dt = dtype_in = np.dtype(dtype_in)
```

Comment:
This function is sometimes called recursively and sometimes the dtype_in is "int" (or another native python type).
The else enforces a conversion to numpy dtype. I'll update the docstring.

Reply:
I understand why it might need conversion on the way in. But why do you change `dtype_in` to the result of the conversion? And why do you use e.g. `dtype_in.names` below?